### ISSUE TITLE
[BUG] - Artstation RSS's appear broken despite being readable in other readers

### ISSUE THREAD

**Issue by Nitrousoxide** [FreshRSS/FreshRSS#5709]
**Describe the bug**
Example RSS feed from Artstation:
https://porforever.artstation.com/rss
It looks to be well formatted and trying to follow it on other readers like ionreader or feeder.co works fine.

Attempting to add it on FreshRSS gives an error:
Blast! The newsfeed could not be added. [Check FreshRSS logs](https://freshrss.internal.(domain name)/i/?a=logs) for details. You can try force adding by appending #force_feed to the URL.

Checking the logs:
`Unknown error for feed [https://porforever.artstation.com/rss]`
--

<br class="Apple-interchange-newline">

**To Reproduce**
1. Go to Add a feed or category
2. Paste in Artstation RSS Feed
3. Click "Add"

**Expected behavior**
The Feed should be added

**Screenshots**

**Environment information (please complete the following information):**
- Device: (Client, Macbook.  Server, Linux)
- OS: [Docker]
- Browser: 118.0.5993.70 Chromium
- FreshRSS version: 1.21.0
- Database version: [SQL Light]
- PHP version: 8.2.10
- Installation type: Docker

**Comment by math-GH** [comment]
I checked it with the current `edge` branch `1.21.1-dev`:

![grafik](https://github.com/FreshRSS/FreshRSS/assets/1645099/0cca866b-8a76-4044-a1d5-315694d75f66)

https://validator.w3.org/feed/check.cgi?url=https%3A%2F%2Fporforever.artstation.com%2Frss
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1645099/9a8ce0ae-85b9-48a5-821d-344137244fdd)

![grafik](https://github.com/FreshRSS/FreshRSS/assets/1645099/980d2849-e6b3-45cc-99bc-86d7ea9b23bd)

Is it maybe a Cloudflare blocker?

**Comment by Nitrousoxide** [comment]
Attempting to curl the url gets this:

```
curl https://porforever.artstation.com/rss
<!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<title>Attention Required! | Cloudflare</title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" /><![endif]-->
<style>body{margin:0;padding:0}</style>


<!--[if gte IE 10]><!-->
<script>
  if (!navigator.cookieEnabled) {
    window.addEventListener('DOMContentLoaded', function () {
      var cookieEl = document.getElementById('cookie-alert');
      cookieEl.style.display = 'block';
    })
  }
</script>
<!--<![endif]-->


</head>
<body>
  <div id="cf-wrapper">
    <div class="cf-alert cf-alert-error cf-cookie-error" id="cookie-alert" data-translate="enable_cookies">Please enable cookies.</div>
    <div id="cf-error-details" class="cf-error-details-wrapper">
      <div class="cf-wrapper cf-header cf-error-overview">
        <h1 data-translate="block_headline">Sorry, you have been blocked</h1>
        <h2 class="cf-subheadline"><span data-translate="unable_to_access">You are unable to access</span> artstation.com</h2>
      </div><!-- /.header -->

      <div class="cf-section cf-highlight">
        <div class="cf-wrapper">
          <div class="cf-screenshot-container cf-screenshot-full">
            
              <span class="cf-no-screenshot error"></span>
            
          </div>
        </div>
      </div><!-- /.captcha-container -->

      <div class="cf-section cf-wrapper">
        <div class="cf-columns two">
          <div class="cf-column">
            <h2 data-translate="blocked_why_headline">Why have I been blocked?</h2>

            <p data-translate="blocked_why_detail">This website is using a security service to protect itself from online attacks. The action you just performed triggered the security solution. There are several actions that could trigger this block including submitting a certain word or phrase, a SQL command or malformed data.</p>
          </div>

          <div class="cf-column">
            <h2 data-translate="blocked_resolve_headline">What can I do to resolve this?</h2>

            <p data-translate="blocked_resolve_detail">You can email the site owner to let them know you were blocked. Please include what you were doing when this page came up and the Cloudflare Ray ID found at the bottom of this page.</p>
          </div>
        </div>
      </div><!-- /.section -->

      <div class="cf-error-footer cf-wrapper w-240 lg:w-full py-10 sm:py-4 sm:px-8 mx-auto text-center sm:text-left border-solid border-0 border-t border-gray-300">
  <p class="text-13">
    <span class="cf-footer-item sm:block sm:mb-1">Cloudflare Ray ID: <strong class="font-semibold">817a66a9186b42b8</strong></span>
    <span class="cf-footer-separator sm:hidden">&bull;</span>
    <span id="cf-footer-item-ip" class="cf-footer-item hidden sm:block sm:mb-1">
      Your IP:
      <button type="button" id="cf-footer-ip-reveal" class="cf-footer-ip-reveal-btn">Click to reveal</button>
      <span class="hidden" id="cf-footer-ip">100.14.15.241</span>
      <span class="cf-footer-separator sm:hidden">&bull;</span>
    </span>
    <span class="cf-footer-item sm:block sm:mb-1"><span>Performance &amp; security by</span> <a rel="noopener noreferrer" href="https://www.cloudflare.com/5xx-error-landing" id="brand_link" target="_blank">Cloudflare</a></span>
    
  </p>
  <script>(function(){function d(){var b=a.getElementById("cf-footer-item-ip"),c=a.getElementById("cf-footer-ip-reveal");b&&"classList"in b&&(b.classList.remove("hidden"),c.addEventListener("click",function(){c.classList.add("hidden");a.getElementById("cf-footer-ip").classList.remove("hidden")}))}var a=document;document.addEventListener&&a.addEventListener("DOMContentLoaded",d)})();</script>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
  </div><!-- /#cf-wrapper -->

  <script>
  window._cf_translation = {};
  
  
</script>

<script>(function(){var js = "window['__CF$cv$params']={r:'817a66a9186b42b8',t:'MTY5NzU2NTIwNS45NDIwMDA='};_cpo=document.createElement('script');_cpo.nonce='',_cpo.src='/cdn-cgi/challenge-platform/scripts/jsd/main.js',document.getElementsByTagName('head')[0].appendChild(_cpo);";var _0xh = document.createElement('iframe');_0xh.height = 1;_0xh.width = 1;_0xh.style.position = 'absolute';_0xh.style.top = 0;_0xh.style.left = 0;_0xh.style.border = 'none';_0xh.style.visibility = 'hidden';document.body.appendChild(_0xh);function handler() {var _0xi = _0xh.contentDocument || _0xh.contentWindow.document;if (_0xi) {var _0xj = _0xi.createElement('script');_0xj.innerHTML = js;_0xi.getElementsByTagName('head')[0].appendChild(_0xj);}}if (document.readyState !== 'loading') {handler();} else if (window.addEventListener) {document.addEventListener('DOMContentLoaded', handler);} else {var prev = document.onreadystatechange || function () {};document.onreadystatechange = function (e) {prev(e);if (document.readyState !== 'loading') {document.onreadystatechange = prev;handler();}};}})();</script></body>
</html>
```

Though get requests (in a browser) go through fine.

<img width="1398" alt="image" src="https://github.com/FreshRSS/FreshRSS/assets/1363931/5dc187f6-099b-429c-9be1-943a89fc8620">

I tried curl'ing with a browser user agent header and still got blocked.  I don't know if there's any other way to get around it.  I presume so since this is supposed to be an rss feed and it'd not be super useful as one if it needed someone to pass a captcha to accessit.

**Comment by Frenzie** [comment]
You could try entering a fake browser user agent in advanced feed settings.

**Comment by math-GH** [comment]
How to do it what @Frenzie told:

![grafik](https://github.com/FreshRSS/FreshRSS/assets/1645099/4a082c40-4582-4224-9a1a-b9abe967644e)

**Comment by Nitrousoxide** [comment]
I'm still getting a 403 error using the example user agent under advanced.
```
2023-10-17 15:55:30 | cURL error 22: The requested URL returned error: 403 [https://porforever.artstation.com/rss]
```

**Comment by Alkarex** [comment]
> Is it maybe a Cloudflare blocker?

Yes, visibly again a case of Cloudflare breaking the Web... It would be interesting to try with https://github.com/lwthiker/curl-impersonate (used e.g. by RSS Bridge) if anyone is motivated to give it a try

**Comment by Nitrousoxide** [comment]
> Yes, visibly again a case of Cloudflare breaking the Web... It would be interesting to try with https://github.com/lwthiker/curl-impersonate (used e.g. by RSS Bridge) if anyone is motivated to give it a try

I did a quick check using their built container and it seemed to work.  Might be a useful addition to freshrss, though I could certainly understand if it's considered out of scope for this project.  curl impersonate itself is only about 30 mb (as a container) so it probably wouldn't add much to the project while solving this (and other) cloudfire breakages.

```
podman run --rm lwthiker/curl-impersonate:0.5-chrome curl_chrome101 https://porforever.artstation.com/rss
Resolving "lwthiker/curl-impersonate" using unqualified-search registries (/etc/containers/registries.conf.d/999-podman-machine.conf)
Trying to pull docker.io/lwthiker/curl-impersonate:0.5-chrome...
Getting image source signatures
Copying blob sha256:59bf1c3509f33515622619af21ed55bbe26d24913cedbca106468a5fb37a50c3
Copying blob sha256:4ff6cca9818448c7d4a16deed3b10f2142f74c0cb5942960402d9d4f35620696
Copying blob sha256:6990413a577969cd8b3e9cfc3cbfcbf5957fbfdbeea8649038b99d4419f3b55a
Copying config sha256:127bc8b8f55418e9779829f1dd863ca7beaff287460e1103d09174c49f7ecd13
Writing manifest to image destination
WARNING: image platform (linux/amd64) does not match the expected platform (linux/arm64)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:content="http://purl.org/rss/1.0/modules/content/">
  <channel>
    <title>Porforever on ArtStation</title>
    <description>Artist</description>
    <link>https://porforever.artstation.com</link>
    <language>en-us</language>
    <atom:link rel="self" type="application/rss+xml" href="https://porforever.artstation.com/rss"/>
    <link href="https://porforever.artstation.com/rss" title="Porforever on ArtStation" type="application/rss+xml" rel="alternate"/>
    <item>
      <title>Middle East Wizards by Porforever</title>
      <description>
        <![CDATA[<p>An attempt to design people of a wizard guild located in the desert land.
<br />2023 © Porforever</p>]]>
      </description>
      <content:encoded>
        <![CDATA[<p>An attempt to design people of a wizard guild located in the desert land.
```

**Comment by Alkarex** [comment]
This is a cat and mouse game with Cloudflare, so I would rather leave this task to a (self-hosted) pass-through service such as [RSS Bridge](https://github.com/RSS-Bridge/rss-bridge), which can just republish invalid feeds (for instance using the built-in *Feed Merge* bridge).
But let's see how it evolves.

Examples through a public instance of RSS Bridge, already not working:
* https://rss-bridge.org/bridge01/?action=display&bridge=FeedMergeBridge&feed_name=Porforever+on+ArtStation&feed_1=https%3A%2F%2Fporforever.artstation.com%2Frss&feed_2=&feed_3=&feed_4=&feed_5=&feed_6=&feed_7=&feed_8=&feed_9=&feed_10=&limit=100&format=Html
* https://rss-bridge.org/bridge01/?action=display&bridge=FeedReducerBridge&url=https%3A%2F%2Fporforever.artstation.com%2Frss&percentage=100&format=Html

Probably due to the IP being blacklisted or something else.

For the record, I have made a little contribution to curl-impersonate https://github.com/lwthiker/curl-impersonate/pull/168

```sh
docker run --rm lwthiker/curl-impersonate:0.6.0-alpha.1-ff-alpine curl_ff109 https://porforever.artstation.com/rss
```

**Comment by Nitrousoxide** [comment]
> This is a cat and mouse game with Cloudflare, so I would rather leave this task to a (self-hosted) pass-through service such as RSS Bridge, which can just republish invalid feeds.

I would note that the issue in this thread is kind of out of scope for RSS Bridge too, since that's a rss feed generator for sites that lack one.  Trying to do pass-through curls for extant rss/atom feeds isn't in their wheel house.

**Comment by Alkarex** [comment]
> I would note that the issue in this thread is kind of out of scope for RSS Bridge too, since that's a rss feed generator for sites that lack one. Trying to do pass-through curls for extant rss/atom feeds isn't in their wheel house.

Indeed, but we fall back to advising RSS Bridge as soon as a Web site does not have a functional RSS feed and isn't simple enough for our built-in XPath approach.

But I am not against a PR to play a bit with curl-impersonate. Adding it is relatively easy, on the model of
https://github.com/RSS-Bridge/rss-bridge/blob/master/Dockerfile
so we could better evaluate the pros and cons. For now, I am more on the side of not doing that kind on things and leave that to a dedicated service (which can be hosted alongside FreshRSS).

**Comment by Frenzie** [comment]
> I would note that the issue in this thread is kind of out of scope for RSS Bridge too

Only arguably for the highly specific subset of has a feed but it's blocked by CF. For the Patreon Bridge as a semi-random example it's just par for the course.

Edit: incidentally, their bypass is quite simple (but possibly broken already)
https://github.com/RSS-Bridge/rss-bridge/blob/8203196145c30587aa83e0d22c9d1c915356e0c9/bridges/PatreonBridge.php#L265-L298
```php
        /*
         * Accept-Language header and the CURL cipher list are for bypassing the
         * Cloudflare anti-bot protection on the Patreon API. If this ever breaks,
         * here are some other project that also deal with this:
         * https://github.com/mikf/gallery-dl/issues/342
         * https://github.com/daemionfox/patreon-feed/issues/7
         * https://www.patreondevelopers.com/t/api-returning-cloudflare-challenge/2025
         * https://github.com/splitbrain/patreon-rss/issues/4
         */
        $header = [
            'Accept-Language: en-US',
            'Content-Type: application/json'
        ];
```

**Comment by 5onderling** [comment]
I found a workaround. If you do https://artstation.com/porforever.rss instead of https://porforever.artstation.com/rss it seems to work, somehow. That is of course only a artstation specific solution.
